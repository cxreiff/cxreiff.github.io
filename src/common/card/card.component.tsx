import { ReactNode } from "react";
import cn from "classnames";

import styles from "./card.module.scss";

export type CardProps = {
  svg?: string;
  primary?: string;
  secondary?: ReactNode;
  tertiary?: ReactNode;
  onClick?: () => void;
  className?: string;
};

export const Card = ({
  svg,
  primary,
  secondary,
  tertiary,
  onClick,
  className,
}: CardProps) => (
  <article
    className={cn(className, styles.card, { [styles.card_clickable]: onClick })}
    onClick={onClick}
  >
    {svg && (
      <div
        className={styles.svg}
        style={{ mask: `url(${svg}) no-repeat center / contain` }}
      />
    )}
    {(primary || secondary) && (
      <div className={styles.title}>
        {primary && (
          <header>
            <h4>{primary}</h4>
          </header>
        )}
        {secondary && <footer>{secondary}</footer>}
      </div>
    )}
    {tertiary && <p>{tertiary}</p>}
  </article>
);
