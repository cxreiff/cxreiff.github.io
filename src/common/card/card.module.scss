@use '../../shared.scss' as *;

.card {
    @include wide_card($accent-color-lighter, $accent-color-darker);
    display: flex;
    flex-wrap: wrap;

    img {
        width: 100%;
        margin: 0 0 1rem 0;
    }

    header {
        flex: 1 0;
        font-size: 1.2rem;
        
        h4 {
            margin: 0.5rem 0;
        }
    }

    footer {
        flex: 0 0 auto;
        font-size: 0.9rem;
        padding: 0.5rem;
    }

    p {
        flex: 0 0 100%;
        font-size: 0.9rem;
        border-top: 2px solid #BDBDBD;
        margin-top: 0;
        padding-top: 1rem;
    }
}

.card_clickable {
    @include interactive(
        0.4rem,
        $accent-color-darker,
        $accent-color-darkest,
        $accent-color-darker,
        $accent-color-darkest
    );
}
