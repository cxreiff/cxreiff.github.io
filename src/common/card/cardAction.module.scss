@use "../../shared.scss" as *;
@use "../../themed.scss" as *;

.card_action {
  padding: 0.5rem 0;

  &.match_height {
    height: calc(100% - 1rem);

    & > div:first-child {
      flex: 1;
      margin: auto;

      @include is_larger_than($sm-breakpoint) {
        margin: 1rem;
      }
    }

    & > div:nth-child(2) {
      padding: 0;
    }

    & header > h4 {
      margin: auto;

      @include is_larger_than($sm-breakpoint) {
        margin: 0.5rem 0;
      }
    }
  }

  @include themed() {
    box-shadow: 0 0 0 2px rgba(t($color-quaternary), 0.2);
  }

  & header > h4 {
    text-align: center;
    font-size: 1rem;
  }
}
