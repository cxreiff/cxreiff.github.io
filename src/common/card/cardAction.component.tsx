import cn from "classnames";

import { Card, CardProps } from "./card.component";

import styles from "./cardAction.module.scss";

type CardActionProps = {
  matchHeight?: boolean;
} & Omit<CardProps, "secondary" | "tertiary">;

export const CardAction = ({
  matchHeight = false,
  className,
  ...props
}: CardActionProps) => (
  <Card
    className={cn(className, styles.card_action, {
      [styles.match_height]: matchHeight,
    })}
    {...props}
  />
);
