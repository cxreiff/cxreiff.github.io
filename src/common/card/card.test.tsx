import { render, fireEvent } from '@testing-library/preact'

import { Card } from './card.component'

test('should render', () => {
    const { container, getAllByText } = render(
        <Card
            image={'test'}
            primary={'test'}
            secondary={'test'}
            tertiary={'test'}
        />
    )

    expect(getAllByText('test')).toHaveLength(3)
    expect(container).toMatchSnapshot()
})

test('should call provided callback', () => {
    const mockHandler = jest.fn()
    const { getAllByText } = render(
        <Card
            image={'test'}
            primary={'test'}
            secondary={'test'}
            tertiary={'test'}
            onClick={mockHandler}
        />
    )

    fireEvent.click(getAllByText('test')[0])
    
    expect(mockHandler).toHaveBeenCalled()
})
