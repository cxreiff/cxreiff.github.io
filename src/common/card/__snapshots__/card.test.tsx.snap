// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render 1`] = `
<div>
  <article
    class="_card_9e273c"
  >
    <div
      class="_svg_9e273c"
      style="mask: url(test) no-repeat center / contain;"
    />
    <div
      class="_title_9e273c"
    >
      <header>
        <h4>
          test
        </h4>
      </header>
      <footer>
        test
      </footer>
    </div>
    <p>
      test
    </p>
  </article>
</div>
`;
